## Names

Names are given to values in Python using an assignment statement.

Let's execute the following statements to get a better understanding!✨

In [5]:
#Hit Shift+Enter to run the cell
a = 10
b = 20
a+b

30

In [4]:
quarter = 1/4
half = 2 * quarter
half

0.5

Please note that, only the current value of an expression is assigned to a name. If that value changes later, names that were defined in terms of that value will not change automatically.

In [5]:
quarter = 4
half

0.5

You will notice that the value of ```half``` doesn't change. 

 - Names must start with a letter, but can contain both letters and numbers.
 - A name cannot contain a space; instead, it is common to use an underscore character ```_``` to replace each space.
 
Typically, more meaningful names can be invented than ```a``` and ```b```. For example, to describe the sales tax on a $5 purchase in Berkeley, CA, the following names clarify the meaning of the various quantities involved.

In [6]:
purchase_price = 5
state_tax_rate = 0.075
county_tax_rate = 0.02
city_tax_rate = 0
sales_tax_rate = state_tax_rate + county_tax_rate + city_tax_rate
sales_tax = purchase_price * sales_tax_rate
sales_tax

0.475

## Growth

The relationship between two measurements of the same quantity taken at different times is often expressed as a growth rate. For example, the United States federal government employed 2,766,000 people in 2002 and 2,814,000 people in 2012. To compute a growth rate, we must first decide which value to treat as the initial amount. For values over time, the earlier value is a natural choice. Then, we divide the difference between the ```changed``` and ```initial``` amount by the ```initial``` amount. 📈

In [7]:
initial = 2766000
changed = 2814000
(changed - initial) / initial

0.01735357917570499

Subtracting one from the ratio of the two measurements also yields the same value.

In [8]:
(changed/initial) - 1

0.017353579175704903

This value is the growth rate over 10 years. A useful property of growth rates is that they don't change even if the values are expressed in different units. So, for example, we can express the same relationship between thousands of people in 2002 and 2012.

In [9]:
initial = 2766
changed = 2814
(changed/initial) - 1

0.017353579175704903

In 10 years, the number of employees of the US Federal Government has increased by only 1.74%. In that time, the total expenditures of the US Federal Government increased from \\$2.37 trillion to \\$3.38 trillion in 2012.

In [10]:
initial = 2.37
changed = 3.38
(changed/initial) - 1

0.4261603375527425

A 42.6% increase in the federal budget is much larger than the 1.74% increase in federal employees. In fact, the number of federal employees has grown much more slowly than the population of the United States, which increased 9.21% in the same time period from 287.6 million people in 2002 to 314.1 million in 2012.

In [11]:
initial = 287.6
changed = 314.1
(changed/initial) - 1

0.09214186369958277

A growth rate can also be negative, representing a decrease in some value. For example, the number of manufacturing jobs in the US decreased from 15.3 million in 2002 to 11.9 million in 2012, a -22.2% growth rate.

Let's execute the following code to find out!

In [8]:
initial = 15.3
changed = 11.9
(changed/initial) - 1

-0.2222222222222222

An annual growth rate is a growth rate of some quantity over a single year. An annual growth rate of 0.035, accumulated each year for 10 years, gives a much larger ten-year growth rate of 0.41 (or 41%).

In [ ]:
1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 - 1

This same computation can be expressed using names and exponents.

In [12]:
annual_growth_rate = 0.035
ten_year_growth_rate = (1 + annual_growth_rate) ** 10 - 1
ten_year_growth_rate

0.410598760621121

Likewise, a ten-year growth rate can be used to compute an equivalent annual growth rate. Below, ```t``` is the number of years that have passed between measurements. The following computes the annual growth rate of federal expenditures over the last 10 years.

In [ ]:
initial = 2.37
changed = 3.38
t = 10
(changed/initial) ** (1/t) - 1

The total growth over 10 years is equivalent to a 3.6% increase each year.

In summary, a growth rate ```g``` is used to describe the relative size of an ```initial``` amount and a ```changed``` amount after some amount of time ```t```. To compute ```changed```, apply the growth rate ```g``` repeatedly, ```t``` times using exponentiation.


```initial * (1 + g) ** t```


To compute ```g```, raise the total growth to the power of ```1/t``` and subtract one.


```(changed/initial) ** (1/t) - 1```

# Congratulations! You have cleared the second milestone!🎊🎉
<img src="Way_to_go.gif" width="750" align="center">